middlewares/authenticator.go

```diff
@@ -76,9 +88,20 @@ func parserBasicUsers(users types.Users) (map[string]string, error) {
 	return userMap, nil
 }
 
-func parserDigestUsers(users types.Users) (map[string]string, error) {
+func parserDigestUsers(digest *types.Digest) (map[string]string, error) {
+	var userStrs []string
+	if digest.UsersFile != "" {
+		var err error
+		if userStrs, err = getUserStringsFromFile(digest.UsersFile); err != nil {
+			return nil, err
+		}
+	}
+	userStrs = append(digest.Users, userStrs...)
 	userMap := make(map[string]string)
-	for _, user := range users {
+	for _, user := range userStrs {
+		if user == "" {
+			continue
+		}
```

Comment:
At first sight, lines [69 to 81](https://github.com/krancour/traefik/blob/a95c9ecbe7bd62cf1e11746e3c66fd1a2c38c39e/middlewares/authenticator.go#L69-L81) and [92 to 104](https://github.com/krancour/traefik/blob/a95c9ecbe7bd62cf1e11746e3c66fd1a2c38c39e/middlewares/authenticator.go#L92-L104) seem to match up quite well. On a second look, the reason why we seemingly cannot extract these parts into a common function (possibly `getUserStringsFromFile()`) easily is because we use two different types (`types.Basic` and `types.Digest`) that are structurally equivalent. While I can't see why they have to be distinct, I admit it might be better to keep it this way for now (mostly because it's not clear to me what the struct field tags do exactly).

We could still extend `getUserStringsFromFile()` to filter out empty strings. I'll leave the decision up to you.